maggy/core/executors/tf_dist_executor.py

```diff
@@ -0,0 +1,370 @@
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+
+import builtins as __builtin__
+import inspect
+import json
+import traceback
+import os
+import random
+import socket
+from typing import Callable, Any, Tuple
+
+import numpy as np
+import tensorflow
+import tensorflow as tf
+
+from maggy import util
+from maggy.core.tf_patching.tf_modules import get_wrapped_model
+from maggy.experiment_config.tf_distributed import TfDistributedConfig
+from maggy.core.rpc import Client
+from maggy.core.reporter import Reporter
+from maggy.core.environment.singleton import EnvSing
+
+
+def dist_executor_fn(
+    train_fn: Callable,
+    config: TfDistributedConfig,
+    num_executors: int,
+    app_id: int,
+    run_id: int,
+    server_addr: str,
+    hb_interval: int,
+    secret: str,
+    log_dir: str,
+) -> Callable:
+    """Wraps the user supplied training function in order to be passed to the Spark Executors.
+
+    :param train_fn: Original training function.
+    :param config: Experiment config.
+    :param app_id: Maggy application ID.
+    :param run_id: Maggy run ID.
+    :param server_addr: IP of the Maggy worker registration RPC server.
+    :param hb_interval: Worker heartbeat interval.
+    :param secret: Secret string to authenticate messages.
+    :param log_dir: Location of the logger file directory on the file system.
+
+    :returns: Patched function to execute on the Spark executors.
+    """
+
+    def wrapper_function(_: Any) -> None:
+        """Patched function from tf_dist_executor_fn factory.
+
+        :param _: Necessary catch for the iterator given by Spark to the
+        function upon foreach calls. Can safely be disregarded.
+        """
+
+        EnvSing.get_instance().set_ml_id(app_id, run_id)
+        partition_id, _ = util.get_partition_attempt_id()
+        client = Client(server_addr, partition_id, 0, hb_interval, secret)
+        log_file = log_dir + "/executor_" + str(partition_id) + ".log"
+
+        reporter = Reporter(log_file, partition_id, 0, __builtin__.print)
+        builtin_print = __builtin__.print
+
+        def maggy_print(*args, **kwargs):
+            builtin_print(*args, **kwargs)
+            reporter.log(" ".join(str(x) for x in args), True)
+
+        __builtin__.print = maggy_print
+
+        try:
+            _register_with_servers(client, reporter, partition_id)
+            tb_logdir, trial_log_file = _setup_logging(reporter, log_dir)
+            reporter.log("Awaiting worker reservations.")
+            client.await_reservations()
+            reporter.log("Reservations complete, configuring Tensorflow.")
+            master_config = client.get_tf_config()
+            reservations = client.get_reservations()
+            if not master_config:
+                reporter.log("Tensorflow registration failed, exiting from all tasks.")
+                return
```

Comment:
I don't understand why you need master_config, but then don't use it anywhere else? what's different to the reservations?